automation/workflow/store.go

```diff
@@ -0,0 +1,305 @@
+		Triggers:  triggers,
+		Hooks:     hooks,
+	}
+	m.mu.Lock()
+	m.workflows[wf.ID] = wf
+	m.mu.Unlock()
+	return wf, nil
+}
+
+// Get fetches a Workflow using the associated ID
+func (m *MemStore) Get(wid ID) (*Workflow, error) {
+	m.mu.Lock()
+	wf, ok := m.workflows[wid]
+	m.mu.Unlock()
+	if !ok {
+		return nil, ErrNotFound
+	}
+	return wf, nil
+}
+
+// GetDatasetWorkflow fetches a workflow using the dataset ID
+func (m *MemStore) GetDatasetWorkflow(did string) (*Workflow, error) {
+	if did == "" {
+		return nil, ErrNotFound
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, wf := range m.workflows {
+		if wf.DatasetID == did {
+			return wf, nil
+		}
+	}
+	return nil, ErrNotFound
+}
+
+// Update updates a workflow with changes
+func (m *MemStore) Update(wf *Workflow) error {
+	if wf == nil {
+		return ErrNilWorkflow
+	}
+	if _, err := m.Get(wf.ID); errors.Is(err, ErrNotFound) {
+		return ErrNotFound
+	}
+	m.mu.Lock()
+	m.workflows[wf.ID] = wf
+	m.mu.Unlock()
+	return nil
+}
+
+// Remove removes a workflow from a store
+func (m *MemStore) Remove(id ID) error {
+	m.mu.Lock()
+	_, ok := m.workflows[id]
+	if !ok {
+		return ErrNotFound
+	}
+	delete(m.workflows, id)
+	m.mu.Unlock()
+	return nil
+}
+
+// Deploy enables the workflow to be run by an orchestrator
+func (m *MemStore) Deploy(id ID) error {
+	wf, err := m.Get(id)
+	if err != nil {
+		return err
+	}
+	wf.Deployed = true
+	return m.Update(wf)
+}
+
+// Undeploy prevents the workflow from being run by an orchestrator
+func (m *MemStore) Undeploy(id ID) error {
+	wf, err := m.Get(id)
+	if err != nil {
+		return err
+	}
+	wf.Deployed = false
+	return m.Update(wf)
+}
+
+// List lists all the workflows in the store, by decending order from time of
+// creation
+func (m *MemStore) List(ctx context.Context, limit, offset int) ([]*Workflow, error) {
+	wfs := NewWorkflowSet()
+	fetchAll := false
+	switch {
+	case limit == -1 && offset == 0:
+		fetchAll = true
+	case limit < 0:
+		return nil, fmt.Errorf("limit of %d is out of bounds", limit)
+	case offset < 0:
+		return nil, fmt.Errorf("offset of %d is out of bounds", offset)
+	case limit == 0:
+		return []*Workflow{}, nil
+	}
+	m.mu.Lock()
+	defer m.mu.Unlock()
+	for _, wf := range m.workflows {
+		wfs.Add(wf)
+	}
+
+	if fetchAll {
+		limit = wfs.Len()
+	}
+	if offset > wfs.Len() {
+		return nil, fmt.Errorf("offset of %d is out of bounds", offset)
+	}
+	sort.Sort(wfs)
+	if offset+limit > wfs.Len() {
+		limit = wfs.Len()
+	}
+	return wfs.Slice(offset, limit), nil
+}
+
+// ListDeployed lists all the workflows in the store that are deployed, by
+// decending order from time of creation
+func (m *MemStore) ListDeployed(ctx context.Context, limit, offset int) ([]*Workflow, error) {
+	wfs := NewWorkflowSet()
+	fetchAll := false
+	switch {
+	case limit == -1 && offset == 0:
+		fetchAll = true
+	case limit < 0:
+		return nil, fmt.Errorf("limit of %d is out of bounds", limit)
+	case offset < 0:
+		return nil, fmt.Errorf("offset of %d is out of bounds", offset)
```

Comment:
Since I there is work happening on listing elsewhere in qri, do we have any standard on what should be returned if the limit and/or offset are negative?